File: clients/go/internal/gate/ask.go
package gate

import (
	"context"
	"errors"
	"math"
	"math/big"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"syscall"
	"time"

	"daisugi-verify/internal/lazyre"
	"daisugi-verify/internal/pyjson"
	"daisugi-verify/internal/pystr"
)

// maybeAsk is gate._maybe_ask: a would-deny is handed to a present
// operator for at most --ask-timeout seconds. With no tool_use_id or no
// present operator the deny stands and nothing is written.
func (r *runner) maybeAsk(p *pyjson.Object, d *decision, sessionID any) *decision {
	tidV := p.Value("tool_use_id")
	if !pyjson.Truthy(tidV) || !r.operatorPresent() {
		return d
	}
	if math.IsNaN(r.askTimeoutS) || math.Abs(r.askTimeoutS) >= 9e18 {
		// int(timeout_s) raises on NaN and the infinities after the wait.
		unported("an --ask-timeout the port does not wait out")
	}
	tid := pyStrOf(tidV)
	deadline := pyTime(time.Now()) + r.askTimeoutS
	r.postAsk(tid, pyjson.NewObject().
		Set("sessionId", p.Value("session_id")).
		Set("toolName", d.ToolName).
		Set("detail", d.Detail).
		Set("reason", d.Reason).
		Set("clause", d.clause()).
		Set("counterexample", d.counterexample()).
		Set("toolInput", p.Value("tool_input")).
		Set("tier", d.Tier), deadline)
	_ = catch(func() { r.reportBlocked(p, d, sessionID, tid, deadline) })
	remaining := math.Max(0, deadline-pyTime(time.Now()))
	reply := r.waitAnswer(tid, remaining)
	out := *d
	var by, whoFrom any
	if reply != nil {
		b, w := whoOf(reply.Value("by"), reply.Value("whoFrom"))
		by, whoFrom = b, w
	}
	out.AnsweredBy, out.WhoFrom = by, whoFrom
	if reply != nil && pyEq(reply.Value("decision"), "allow") {
		why := reply.Value("reason")
		if !pyjson.Truthy(why) {
			why = "no reason given"
		}
		if pyEq(reply.Value("scope"), "task") && d.Tier == tierUndoable {
			r.proposeForTask(p, d, tid)
		}
		out.Allow = true
		out.Ask = true
		out.Reason = "allowed by operator: " + pyStrOf(why)
		out.UpdatedInput = nil
		if u := reply.Value("updatedInput"); pyjson.Truthy(u) {
			out.UpdatedInput = u
		}
		return &out
	}
	why := "operator did not answer within " + strconv.FormatInt(int64(r.askTimeoutS), 10) + " s"
	if reply != nil {
		why = "operator denied"
	}
	out.Ask = reply != nil
	out.Reason = d.Reason + " (" + why + ")"
	return &out
}

// askSafe is ask._safe.
func askSafe(raw any) string {
	s := ""
	if pyjson.Truthy(raw) {
		s = pyStrOf(raw)
	}
	var b strings.Builder
	for _, c := range pystr.Runes(s) {
		if c < 0x80 && (c == '.' || c == '_' || c == '-' || ('0' <= c && c <= '9') || ('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z')) {
			b.WriteRune(c)
		} else {
			b.WriteByte('_')
		}
	}
	out := pyHead(strings.Trim(b.String(), "."), 128)
	if out == "" {
		return "none"
	}
	return out
}

// askMkdir is ask._mkdir.
func askMkdir(dir string) {
	enc := fsEnc(dir)
	if err := os.MkdirAll(filepath.Dir(enc), 0o777); err != nil {
		panic(osError(err, dir))
	}
	if err := os.Mkdir(enc, 0o700); err != nil {
		if st, serr := os.Stat(enc); serr != nil || !st.IsDir() {
			panic(osError(err, dir))
		}
	}
	_ = os.Chmod(enc, 0o700)
}

func fsEnc(p string) string {
	b, err := pystr.FSEncode(p)
	if err != nil {
		panic(err)
	}
	return string(b)
}

// askWriteJSON is ask._write_json: a temp file beside the target, then a
// rename, so a reader never sees half a file.
func askWriteJSON(path string, body *pyjson.Object) {
	askMkdir(pathParent(path))
	name := path[strings.LastIndex(path, "/")+1:]
	stem := name
	if i := strings.LastIndex(name, "."); i > 0 {
		stem = name[:i]
	}
	tmp := pathJoin(pathParent(path), stem+".tmp")
	if err := os.WriteFile(fsEnc(tmp), []byte(pyjson.Dumps(body, true)), 0o666); err != nil {
		panic(osError(err, tmp))
	}
	_ = os.Chmod(fsEnc(tmp), 0o600)
	if err := os.Rename(fsEnc(tmp), fsEnc(path)); err != nil {
		panic(osError(err, tmp))
	}
}

// unlinkMissingOK is Path.unlink with FileNotFoundError caught: any other
// error raises.
func unlinkMissingOK(p string) {
	if err := syscall.Unlink(fsEnc(p)); err != nil && err != syscall.ENOENT {
		panic(osError(err, p))
	}
}

// globJSON is Path(dir).glob("*.json"): every entry whose name ends in
// .json, hidden ones included, in directory order.
func globJSON(dir string) []string {
	ents, err := os.ReadDir(fsEnc(dir))
	if err != nil {
		return nil
	}
	var out []string
	for _, e := range ents {
		if n := pystr.FSDecode([]byte(e.Name())); strings.HasSuffix(n, ".json") {
			out = append(out, n)
		}
	}
	return out
}

// sweepExpired is ask._sweep_expired.
func (r *runner) sweepExpired(now float64) {
	asks, answers := pathJoin(r.root, "asks"), pathJoin(r.root, "answers")
	if exists(asks) {
		for _, name := range globJSON(asks) {
			p := pathJoin(asks, name)
			body := readJSONObject(p)
			deadline := 0.0
			if body != nil {
				if v, ok := body.Get("deadline"); ok {
					if f, ok := pyFloatOf(v); ok {
						deadline = f
					}
				}
			}
			if body == nil || deadline <= now {
				unlinkMissingOK(p)
				unlinkMissingOK(pathJoin(answers, name))
			}
		}
	}
	if exists(answers) {
		for _, name := range globJSON(answers) {
			if !exists(pathJoin(asks, name)) {
				unlinkMissingOK(pathJoin(answers, name))
			}
		}
	}
}

// pyFloatOf is float(v) for a JSON value; false where Python raises.
func pyFloatOf(v any) (float64, bool) {
	switch x := v.(type) {
	case bool:
		if x {
			return 1, true
		}
		return 0, true
	case pyjson.Int:
		f, ok := new(big.Float).SetString(x.Text)
		if !ok {
			return 0, false
		}
		out, acc := f.Float64()
		if math.IsInf(out, 0) && acc != big.Exact {
			// float() of an int past the float range raises OverflowError.
			return 0, false
		}
		return out, true
	case pyjson.Float:
		return float64(x), true
	case string:
		f := pyFloat(x)
		if f == nil {
			return 0, false
		}
		return *f, true
	}
	return 0, false
}

// postAsk is ask.post_ask.
func (r *runner) postAsk(tid string, question *pyjson.Object, deadline float64) {
	r.sweepExpired(pyTime(time.Now()))
	body := pyjson.NewObject().
		Set("toolUseId", tid).
		Set("nonce", randomHex(16)).
		Set("postedAt", pyjson.Float(pyTime(time.Now()))).
		Set("deadline", pyjson.Float(deadline))
	for _, k := range question.Keys() {
		body.Set(k, question.Value(k))
	}
	askWriteJSON(pathJoin(pathJoin(r.root, "asks"), askSafe(tid)+".json"), body)
}

var (
	whoName   = lazyre.New(`^[A-Za-z0-9_-]{1,32}$`)
	whoPane   = lazyre.New(`^pane(:[A-Za-z0-9._:-]{1,128})?$`)
	whoPlugin = lazyre.New(`^plugin:[A-Za-z0-9._:-]{1,128}$`)
)

// whoOf is ask.who_of.
func whoOf(by, whoFrom any) (string, string) {
	w, ok1 := whoFrom.(string)
	b, ok2 := by.(string)
	if !ok1 || !ok2 {
		return "local", "none"
	}
	var rule *regexp.Regexp
	switch w {
	case "token", "socket":
		rule = whoName()
	case "pane":
		rule = whoPane()
	case "plugin":
		rule = whoPlugin()
	}
	if rule == nil || !rule.MatchString(b) {
		return "local", "none"
	}
	return b, w
}

// consumeAsk is ask._consume.
func (r *runner) consumeAsk(tid string) {
	name := askSafe(tid) + ".json"
	unlinkMissingOK(pathJoin(pathJoin(r.root, "asks"), name))
	unlinkMissingOK(pathJoin(pathJoin(r.root, "answers"), name))
}

func mtime(p string) (float64, bool) {
	st, err := os.Stat(fsEnc(p))
	if err != nil {
		return 0, false
	}
	return float64(st.ModTime().UnixNano()) / 1e9, true
}

// waitAnswer is ask.wait_answer: an answer is honored only when it names
// allow or deny, echoes the ask's nonce and is no older than the ask.
func (r *runner) waitAnswer(tid string, timeoutS float64) *pyjson.Object {
	name := askSafe(tid) + ".json"
	askPath := pathJoin(pathJoin(r.root, "asks"), name)
	answerPath := pathJoin(pathJoin(r.root, "answers"), name)
	askBody := readJSONObject(askPath)
	var nonce any
	if askBody != nil {
		nonce = askBody.Value("nonce")
	}
	askM, askOK := mtime(askPath)
	end := time.Now().Add(time.Duration(timeoutS * float64(time.Second)))
	for {
		if exists(answerPath) {
			body := readJSONObject(answerPath)
			ansM, ansOK := mtime(answerPath)
			valid := body != nil &&
				(pyEq(body.Value("decision"), "allow") || pyEq(body.Value("decision"), "deny")) &&
				nonce != nil && pyEq(body.Value("nonce"), nonce) &&
				askOK && ansOK && ansM >= askM
			r.consumeAsk(tid)
			if !valid {
				return nil
			}
			out := pyjson.NewObject()
			for _, k := range body.Keys() {
				if k != "nonce" {
					out.Set(k, body.Value(k))
				}
			}
			return out
		}
		if !time.Now().Before(end) {
			r.consumeAsk(tid)
			return nil
		}
		time.Sleep(200 * time.Millisecond)
	}
}

// proposeForTask is gate._propose_for_task: best effort.
func (r *runner) proposeForTask(p *pyjson.Object, d *decision, tid string) {
	_ = catch(func() {
		id := randomHex(4)
		now := pyTime(time.Now())
		askWriteJSON(pathJoin(pathJoin(r.root, "proposals"), id+".json"), pyjson.NewObject().
			Set("id", id).
			Set("kind", "allow-pattern").
			Set("scope", "task").
			Set("expiresAt", pyjson.Float(now+30*86400)).
			Set("createdAt", pyjson.Float(pyTime(time.Now()))).
			Set("sessionId", p.Value("session_id")).
			Set("toolUseId", tid).
			Set("toolInput", p.Value("tool_input")).
			Set("clause", d.clause()).
			Set("tier", d.Tier))
	})
}

// readJSONObject is ask._read_json: the file's JSON object, or nil for a
// file that cannot be read, is not JSON, or is not an object.
func readJSONObject(p string) *pyjson.Object {
	enc, eerr := pystr.FSEncode(p)
	if eerr != nil {
		panic(eerr)
	}
	raw, err := os.ReadFile(string(enc))
	if err != nil {
		return nil
	}
	text, derr := pystr.DecodeStrict(raw)
	if derr != nil {
		return nil
	}
	v, jerr := pyjson.Loads(text)
	if jerr != nil {
		return nil
	}
	o, _ := v.(*pyjson.Object)
	return o
}

// operatorPresent is ask.operator_present: a heartbeat file under 15 s old
// naming a live pid.
func (r *runner) operatorPresent() bool {
	p := pathJoin(r.root, "operator.json")
	body := readJSONObject(p)
	if body == nil || body.Len() == 0 {
		return false
	}
	enc, _ := pystr.FSEncode(p)
	st, err := os.Stat(string(enc))
	if err != nil {
		return false
	}
	age := float64(time.Now().UnixNano())/1e9 - float64(st.ModTime().UnixNano())/1e9
	if age > 15 {
		return false
	}
	var pid int64
	switch x := body.Value("pid").(type) {
	case bool:
		if x {
			pid = 1
		}
	case pyjson.Int:
		n, ok := new(big.Int).SetString(x.Text, 10)
		if !ok || !n.IsInt64() || n.Int64() < -1<<31 || n.Int64() >= 1<<31 {
			// os.kill raises OverflowError past a C int.
			unported("an operator pid past a C int")
		}
		pid = n.Int64()
	default:
		return false
	}
	err = syscall.Kill(int(pid), 0)
	switch {
	case err == nil, errors.Is(err, syscall.EPERM):
		return true
	case errors.Is(err, syscall.ESRCH):
		return false
	}
	panic(osError(err, ""))
}

// isRepo is checkpoints.is_repo: git rev-parse --is-inside-work-tree,
// hardened the same way every call in checkpoint.go is (gitSafeArgs, the
// two GIT_CONFIG_* overrides, the checkpoint's own shared deadline). This
// runs in the same WORKSPACE repo a checkpoint's other git calls do,
// before checkpoints.py even knows whether one is due, and it runs twice
// (dueCheckpoint, then snapshot), so it must not carry a timeout of its
// own outside that shared budget.
func (r *runner) isRepo(dir string) bool {
	enc, eerr := pystr.FSEncode(dir)
	if eerr != nil {
		panic(eerr)
	}
	ctx, cancel := context.WithDeadline(context.Background(), r.gitDeadline())
	defer cancel()
	argv := append(append([]string{"-C", string(enc)}, gitSafeArgs...), "rev-parse", "--is-inside-work-tree")
	cmd := exec.CommandContext(ctx, "git", argv...)
	cmd.Env = append(r.environ(), "GIT_CONFIG_NOSYSTEM=1", "GIT_CONFIG_GLOBAL=/dev/null")
	out, err := cmd.Output()
	if err != nil {
		return false
	}
	text, derr := pystr.DecodeStrict(out)
	if derr != nil {
		panic(derr)
	}
	return strings.TrimRight(universalNewlines(text), "\n") == "true"
}

File: clients/rust/src/gate/ask.rs
//! `gate._maybe_ask` and the file protocol of `opendaisugi.ask`: a
//! would-deny is handed to a present operator for at most --ask-timeout.
//! No operator, no tool_use_id, a deny, a late or a forged answer all
//! leave the deny standing; only an answer that echoes this ask's nonce
//! and is no older than it can allow.

use super::decide::Decision;
use super::effects::UNDOABLE;
use super::paths::{os_error, os_path};
use super::pymodel::decode_utf8_strict;
use super::pyjson::{dumps, loads, py_str, LoadError, Object, Value};
use super::{random_hex, undecided, Fault, PyErr, Runner, R};
use std::os::unix::fs::{DirBuilderExt, PermissionsExt};
use std::path::{Path, PathBuf};
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

const PRESENCE: &str = "operator.json";
const ASKS: &str = "asks";
const ANSWERS: &str = "answers";
const PROPOSALS: &str = "proposals";

/// `ask._safe(raw)`: `str(raw or "")` with every character but ASCII
/// letters, digits, `.`, `_` and `-` made `_`, dots stripped from the
/// ends, at most 128 characters, or `none`.
fn safe(raw: &str) -> String {
    let s: String =
        raw.chars().map(|c| if c.is_ascii_alphanumeric() || matches!(c, '.' | '_' | '-') { c } else { '_' }).collect();
    let s: String = s.trim_matches('.').chars().take(128).collect();
    if s.is_empty() {
        "none".into()
    } else {
        s
    }
}

fn io(e: std::io::Error, p: &Path) -> Fault {
    Fault::Raised(os_error(&e, &p.to_string_lossy()))
}

/// `ask._mkdir`.
fn mkdir(p: &Path) -> R<()> {
    std::fs::DirBuilder::new().recursive(true).mode(0o700).create(p).map_err(|e| io(e, p))?;
    let _ = std::fs::set_permissions(p, std::fs::Permissions::from_mode(0o700));
    Ok(())
}

/// `ask._write_json(path, body)`: through a `.tmp` file and `os.replace`.
fn write_json(path: &Path, body: &Object) -> R<()> {
    if let Some(parent) = path.parent() {
        mkdir(parent)?;
    }
    let tmp = path.with_extension("tmp");
    std::fs::write(&tmp, dumps(&Value::Obj(body.clone()), true)).map_err(|e| io(e, &tmp))?;
    let _ = std::fs::set_permissions(&tmp, std::fs::Permissions::from_mode(0o600));
    std::fs::rename(&tmp, path).map_err(|e| io(e, path))
}

/// `ask._read_json(path)`: a dict, or None for any OSError or ValueError.
fn read_json(path: &Path) -> R<Option<Object>> {
    let raw = match std::fs::read(path) {
        Ok(r) => r,
        Err(_) => return Ok(None),
    };
    let text = match decode_utf8_strict(&raw) {
        Ok(t) => t,
        Err(_) => return Ok(None),
    };
    match loads(&text) {
        Ok(Value::Obj(o)) => Ok(Some(o)),
        Ok(_) => Ok(None),
        // json.loads raises RecursionError, which is no ValueError.
        Err(LoadError::Recursion) => undecided("an ask file nested past Python's recursion limit"),
        Err(_) => Ok(None),
    }
}

fn mtime(p: &Path) -> Option<f64> {
    let m = std::fs::metadata(p).ok()?.modified().ok()?;
    Some(m.duration_since(UNIX_EPOCH).map(|d| d.as_secs_f64()).unwrap_or(0.0))
}

/// `os.kill(pid, 0)` for `_pid_alive`, with the OverflowError Python's
/// argument conversion raises.
fn pid_alive(pid: &Value) -> R<bool> {
    let n: i128 = match pid {
        Value::Bool(b) => *b as i128,
        Value::Int(t) => match t.parse::<i128>() {
            Ok(n) => n,
            Err(_) => return Err(PyErr::new("OverflowError", "Python int too large to convert to C long").into()),
        },
        _ => return Ok(false),
    };
    if n > i64::MAX as i128 || n < i64::MIN as i128 {
        return Err(PyErr::new("OverflowError", "Python int too large to convert to C long").into());
    }
    if n > i32::MAX as i128 {
        return Err(PyErr::new("OverflowError", "signed integer is greater than maximum").into());
    }
    if n < i32::MIN as i128 {
        return Err(PyErr::new("OverflowError", "signed integer is less than minimum").into());
    }
    let r = unsafe { libc::kill(n as i32, 0) };
    if r == 0 {
        return Ok(true);
    }
    let errno = std::io::Error::last_os_error().raw_os_error().unwrap_or(0);
    Ok(errno != libc::ESRCH)
}

/// Python's `float(x)` for an ask's deadline: `None` for the TypeError or
/// ValueError `_sweep_expired` catches.
fn py_float(v: &Value) -> R<Option<f64>> {
    Ok(match v {
        Value::Bool(b) => Some(if *b { 1.0 } else { 0.0 }),
        Value::Int(t) => {
            let f: f64 = t.parse().unwrap_or(f64::INFINITY);
            if f.is_infinite() {
                return Err(PyErr::new("OverflowError", "int too large to convert to float").into());
            }
            Some(f)
        }
        Value::Float(f) => Some(*f),
        Value::Str(s) => super::argparse::py_float(s),
        _ => None,
    })
}

/// `ask.who_of(by, who_from)`.
pub fn who_of(by: &Value, who_from: &Value) -> (String, String) {
    let local = ("local".to_string(), "none".to_string());
    let (by, wf) = match (by, who_from) {
        (Value::Str(b), Value::Str(w)) => (b.as_str(), w.as_str()),
        _ => return local,
    };
    let charset = |s: &str, extra: &str, max: usize| {
        let n = s.chars().count();
        n >= 1 && n <= max && s.chars().all(|c| c.is_ascii_alphanumeric() || extra.contains(c))
    };
    let ok = match wf {
        "token" | "socket" => charset(by, "_-", 32),
        "pane" => by == "pane" || by.strip_prefix("pane:").is_some_and(|r| charset(r, "._:-", 128)),
        "plugin" => by.strip_prefix("plugin:").is_some_and(|r| charset(r, "._:-", 128)),
        _ => false,
    };
    if ok {
        (by.to_string(), wf.to_string())
    } else {
        local
    }
}

fn now() -> f64 {
    SystemTime::now().duration_since(UNIX_EPOCH).map(|d| d.as_secs_f64()).unwrap_or(0.0)
}

/// `path.unlink()` with FileNotFoundError ignored and any other OSError
/// raised.
fn unlink_missing_ok(p: &Path) -> R<()> {
    match std::fs::remove_file(p) {
        Ok(()) => Ok(()),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(()),
        Err(e) => Err(io(e, p)),
    }
}

/// `Path.glob("*.json")` in one directory: dotfiles included, directories
/// too.
fn json_names(dir: &Path) -> Vec<PathBuf> {
    match std::fs::read_dir(dir) {
        Ok(rd) => rd
            .filter_map(|e| e.ok())
            .map(|e| e.path())
            .filter(|p| p.file_name().and_then(|n| n.to_str()).is_some_and(|n| n.ends_with(".json")))
            .collect(),
        Err(_) => vec![],
    }
}

impl Runner {
    fn gate_root(&self) -> PathBuf {
        match os_path("open", &self.root) {
            Ok(p) => p,
            Err(_) => PathBuf::from(&self.root),
        }
    }

    /// `ask.operator_present(root)`.
    fn operator_present(&self) -> R<bool> {
        let p = self.gate_root().join(PRESENCE);
        let body = match read_json(&p)? {
            Some(b) => b,
            None => return Ok(false),
        };
        let age = match mtime(&p) {
            Some(m) => now() - m,
            None => return Ok(false),
        };
        if age > 15.0 {
            return Ok(false);
        }
        match body.value("pid") {
            v @ (Value::Int(_) | Value::Bool(_)) => pid_alive(v),
            _ => Ok(false),
        }
    }

    /// `ask._sweep_expired(root, now=now)`.
    fn sweep_expired(&self, at: f64) -> R<()> {
        let root = self.gate_root();
        let asks = root.join(ASKS);
        if asks.exists() {
            for p in json_names(&asks) {
                let body = read_json(&p)?;
                let deadline = match &body {
                    Some(b) => match b.get("deadline") {
                        None => 0.0,
                        Some(v) => py_float(v)?.unwrap_or(0.0),
                    },
                    None => 0.0,
                };
                if body.is_none() || deadline <= at {
                    unlink_missing_ok(&p)?;
                    if let Some(name) = p.file_name() {
                        unlink_missing_ok(&root.join(ANSWERS).join(name))?;
                    }
                }
            }
        }
        let answers = root.join(ANSWERS);
        if answers.exists() {
            for p in json_names(&answers) {
                if let Some(name) = p.file_name() {
                    if !asks.join(name).exists() {
                        unlink_missing_ok(&p)?;
                    }
                }
            }
        }
        Ok(())
    }

    /// `ask._consume(root, tool_use_id)`.
    fn consume(&self, tid: &str) -> R<()> {
        let root = self.gate_root();
        for sub in [ASKS, ANSWERS] {
            unlink_missing_ok(&root.join(sub).join(format!("{}.json", safe(tid))))?;
        }
        Ok(())
    }

    /// `ask.wait_answer(root, tool_use_id=tid, timeout_s=...)`.
    fn wait_answer(&self, tid: &str, timeout_s: f64) -> R<Option<Object>> {
        let root = self.gate_root();
        let name = format!("{}.json", safe(tid));
        let ask_path = root.join(ASKS).join(&name);
        let answer_path = root.join(ANSWERS).join(&name);
        let expected = read_json(&ask_path)?.map(|b| b.value("nonce").clone()).unwrap_or(Value::Null);
        let ask_mtime = mtime(&ask_path);
        let t_end = Instant::now() + Duration::from_secs_f64(timeout_s.clamp(0.0, 1e9));
        loop {
            if answer_path.exists() {
                let body = read_json(&answer_path)?;
                let answer_mtime = mtime(&answer_path);
                let valid = match &body {
                    Some(b) => {
                        let decision = b.value("decision");
                        (decision == &Value::Str("allow".into()) || decision == &Value::Str("deny".into()))
                            && !expected.is_null()
                            && super::predicate::py_eq(b.value("nonce"), &expected)
                            && match (ask_mtime, answer_mtime) {
                                (Some(a), Some(b)) => b >= a,
                                _ => false,
                            }
                    }
                    None => false,
                };
                self.consume(tid)?;
                if !valid {
                    return Ok(None);
                }
                let b = body.unwrap_or_default();
                let mut out = Object::new();
                for k in b.keys() {
                    if k != "nonce" {
                        out.set(k, b.value(k).clone());
                    }
                }
                return Ok(Some(out));
            }
            if Instant::now() >= t_end {
                self.consume(tid)?;
                return Ok(None);
            }
            std::thread::sleep(Duration::from_millis(200));
        }
    }

    /// `gate._maybe_ask`.
    pub fn maybe_ask(&self, p: &Object, d: Decision, session_id: &Value) -> R<Decision> {
        let tool_use_id = p.value("tool_use_id");
        if !tool_use_id.truthy() || !self.operator_present()? {
            return Ok(d);
        }
        if !self.ask_timeout_s.is_finite() {
            // int(timeout_s) in the reason raises, and an infinite wait
            // is no answer the port can give.
            return undecided("an --ask-timeout that is not finite");
        }
        let tid = py_str(tool_use_id)?;
        let deadline = now() + self.ask_timeout_s;
        // post_ask
        self.sweep_expired(now())?;
        let tool_name = match &d.tool_name {
            Some(t) => Value::Str(t.clone()),
            None => Value::Null,
        };
        let body = Object::new()
            .with("toolUseId", tid.as_str())
            .with("nonce", random_hex(16)?)
            .with("postedAt", now())
            .with("deadline", deadline)
            .with("sessionId", p.value("session_id").clone())
            .with("toolName", tool_name)
            .with("detail", d.detail.as_str())
            .with("reason", d.reason.as_str())
            .with("clause", d.clause())
            .with("counterexample", d.counterexample())
            .with("toolInput", p.value("tool_input").clone())
            .with("tier", d.tier.as_str());
        write_json(&self.gate_root().join(ASKS).join(format!("{}.json", safe(&tid))), &body)?;
        let _ = super::best_effort(self.report_blocked(p, &d, session_id, &tid, deadline));
        let remaining = (deadline - now()).max(0.0);
        let reply = self.wait_answer(&tid, remaining)?;
        let (by, who_from) = match &reply {
            Some(r) => {
                let (b, w) = who_of(r.value("by"), r.value("whoFrom"));
                (Some(b), Some(w))
            }
            None => (None, None),
        };
        let mut d = d;
        if let Some(r) = &reply {
            if r.value("decision") == &Value::Str("allow".into()) {
                let why = if r.value("reason").truthy() { py_str(r.value("reason"))? } else { "no reason given".into() };
                if r.value("scope") == &Value::Str("task".into()) && d.tier == UNDOABLE {
                    let _ = super::best_effort(self.propose_for_task(p, &d, &tid));
                }
                d.allow = true;
                d.ask = true;
                d.reason = format!("allowed by operator: {why}");
                d.updated_input = if r.value("updatedInput").truthy() { Some(r.value("updatedInput").clone()) } else { None };
                d.answered_by = by;
                d.who_from = who_from;
                return Ok(d);
            }
        }
        let why = if reply.is_some() {
            "operator denied".to_string()
        } else {
            format!("operator did not answer within {} s", self.ask_timeout_s.trunc() as i128)
        };
        d.ask = reply.is_some();
        d.reason = format!("{} ({why})", d.reason);
        d.answered_by = by;
        d.who_from = who_from;
        Ok(d)
    }

    /// `gate._propose_for_task`.
    fn propose_for_task(&self, p: &Object, d: &Decision, tid: &str) -> R<()> {
        let pid = random_hex(4)?;
        let body = Object::new()
            .with("id", pid.as_str())
            .with("kind", "allow-pattern")
            .with("scope", "task")
            .with("expiresAt", now() + 30.0 * 86400.0)
            .with("createdAt", now())
            .with("sessionId", p.value("session_id").clone())
            .with("toolUseId", tid)
            .with("toolInput", p.value("tool_input").clone())
            .with("clause", d.clause())
            .with("tier", d.tier.as_str());
        write_json(&self.gate_root().join(PROPOSALS).join(format!("{pid}.json")), &body)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn who_of_checks_the_name() {
        let s = |x: &str| Value::Str(x.into());
        assert_eq!(who_of(&s("ann"), &s("token")), ("ann".into(), "token".into()));
        assert_eq!(who_of(&s("pane:w1:p2"), &s("pane")), ("pane:w1:p2".into(), "pane".into()));
        assert_eq!(who_of(&s("pane"), &s("pane")), ("pane".into(), "pane".into()));
        assert_eq!(who_of(&s("a b"), &s("token")), ("local".into(), "none".into()));
        assert_eq!(who_of(&s("x"), &s("nobody")), ("local".into(), "none".into()));
        assert_eq!(who_of(&Value::Null, &s("token")), ("local".into(), "none".into()));
    }

    #[test]
    fn safe_names() {
        assert_eq!(safe("toolu_01.x"), "toolu_01.x");
        assert_eq!(safe("..a/b.."), "a_b");
        assert_eq!(safe(""), "none");
        assert_eq!(safe("é"), "_");
    }
}
